### ISSUE TITLE
Multi-line find doesn't honor empty hard returns

### ISSUE THREAD

**Issue by seattledeveloper** [Komodo/KomodoEdit#891]
Hello,

I think this may be a bug, definitely not behavior I was expecting.

When doing a multi-line search, the find functionality finds all items ignoring empty hard returns entered in the find box.  Example:

```text
some text[hard return]
[hard return]
```

returned all instances of `some text`, irregardless of whether or not if was followed by an empty line.  My search was looking for the text followed by an empty line.

I also notice that when I bring up dialog a second time, that the hard returns have been stripped.  It seems that you are removing white space from your searches.  This makes it impossible to search for white space, unless you use regular expressions.

Thanks for your thoughts on this.

**Comment by Naatan** [comment]
I cannot reproduce this. Could you please make a screen recording so we have the full steps to reproduce?

**Comment by seattledeveloper** [comment]
Here's a screen shot showing the search (with the cursor two hard returns down in the find dialog) and showing 2 matches in the search results pane even though only the first result should match the entered search:

![full-screen-shot](https://cloud.githubusercontent.com/assets/9749337/11355411/b4b4042e-9209-11e5-848b-4419fbbda0b2.png)

And here's a screen shot from when I invoke the search dialog a second time, showing the cursor and the end of the first line (the two hard returns have been removed by Komodo):

![find-again](https://cloud.githubusercontent.com/assets/9749337/11355471/05109e00-920a-11e5-807b-74482fadf768.png)

Hope this is helpful.

**Comment by seattledeveloper** [comment]
FYI, I just upgraded to 9.3.1.  The search behavior is the same (it still displays 2 results even though only 1 should match) but the extra hard returns are no longer being removed when the find in files dialog is brought up the second time.

**Comment by Naatan** [comment]
Ahh right, so the issue is not that it is searching without the line endings, the issue is that the find dialog trims the search query when it is restored.

Thanks for explaining!

**Comment by seattledeveloper** [comment]
I believe it's actually that the string is being truncated before the search is performed. My search with 2 hard returns should have only matched on line 1 where there an empty vertical space, and not on line 6 where there is not an empty vertical space.  And in 9.3.1, the find dialog no longer seems to be trimming the string when the search query is restored.

**Comment by Naatan** [comment]
The only part of this that I am able to reproduce (on 9.3.1) is Find not preserving the starting/ending whitespace on restore. The rest works as expected.

Perhaps you could provide a text snippet with a search string, so I have your exact reproduction steps?

**Comment by seattledeveloper** [comment]
Here's my exact search string with the two hard returns -- it's really tough to illustrate them here in this editor

```text
$ret_str .= <<<END


```

**Comment by Naatan** [comment]
Could you please provide the text that you are searching on as well?

**Comment by seattledeveloper** [comment]
[text_being_searched.txt](https://github.com/Komodo/KomodoEdit/files/42937/text_being_searched.txt)
[search_query.txt](https://github.com/Komodo/KomodoEdit/files/42938/search_query.txt)

**Comment by seattledeveloper** [comment]
Ah.  I just figured it out.  It seems I needed to add a second (third actually) hard return to get the search to return correctly. Maybe that is the actual issue, and why you couldn't reproduce this.  I was looking at where the cursor was (you can see it in my screenshot), and was thinking I'd added enough vertical space to get the desired result.  Could it still be a bug that I had to enter 3 hard returns when 2 should show an empty line?

**Comment by Naatan** [comment]
Right, so that is matching both instances, as is to be expected. Note your search query just has one new line marker, which regardless of the next line being empty or not is what is being matched accordingly.

If I add an extra new line marker to your search query then it only matches the first instance.

It seems you may be confusing the new line marker with something that represents an "empty line"?

**Comment by seattledeveloper** [comment]
Actually, it isn't as expected. I just tried a different approach.  I selected the exact text to search on, including the trailing hard return, pasted it into the find in files box, and it returns both lines.  I have to manually *add* an *extra* hard return to have it search correctly.  This is what was throwing me off when I first added this issue.

**Comment by cgchoffman** [comment]
I believe @Naatan is correct.

You're search is as follows (assumes you know `\n` is new line)
`$ret_str .= <<<END\n\n`

The string that you don't think it should match is:
`$ret_str .= <<<END\n\nJOE`

As you can see, it does and should match.

The example that you gave that did what you expected is:
`$ret_str .= <<<END\n\n\n`

Which correctly doesn't match
`$ret_str .= <<<END\n\nJOE`

The find is functioning as expect.

**Comment by seattledeveloper** [comment]
Actually, the one that I don't think should match is
```text
$ret_str .= <<<END\nJOE
```
which is showing up in my search results when I search on
```text
$ret_str .= <<<END\n\n
```
If you look at my original screenshot and where the cursor is in the find in files dialog box, you'll see the cursor two hard returns down, which I assume means I'm searching on this
```text
$ret_str .= <<<END\n\n
```
which I don't believe should match
```text
$ret_str .= <<<END\nJOE
```
which it shows up matching on line 6.

**Comment by Naatan** [comment]
I am unable to reproduce that though, even with the snippet you provided.

**Comment by cgchoffman** [comment]
BAH, You're right @seattledeveloper. My example was wrong...and so long.  I'm able to reproduce this and I was able to created a new example that also reproduces it.

**Comment by Naatan** [comment]
Found the issue, it occurs when you use windows line endings. In actuality your sample snippet is: 

`$ret_str .= <<<END\r\nJOE`

Both \r and \n are matched against the \n given in the search pattern.

So the issue is that we need to preserve the line endings in the search pattern as they were defined in the current file.

**Comment by seattledeveloper** [comment]
Yes.  That's it.  Perfect!  Thanks for you and @cgchoffman bearing with me through the process of identifying the issue.
